libsolidity/analysis/ConstantEvaluator.h

```diff
@@ -43,17 +44,21 @@ class TypeChecker;
 class ConstantEvaluator: private ASTConstVisitor
 {
 public:
-	ConstantEvaluator(
-		langutil::ErrorReporter& _errorReporter,
-		size_t _newDepth = 0,
-		std::shared_ptr<std::map<ASTNode const*, TypePointer>> _types = std::make_shared<std::map<ASTNode const*, TypePointer>>()
-	):
-		m_errorReporter(_errorReporter),
-		m_depth(_newDepth),
-		m_types(std::move(_types))
+	struct TypedValue { TypePointer sourceType; TypePointer evaluatedValue; };
```

Comment:
The idea here is that `sourceType` can be nullptr, right? Also shouldn't it be `RationalNumberType const*` instead of `TypePointer`?

Reply:
`sourceType` defaults to the other type pointer (which can only be `RationalNumberType`). I did thise in order to avoid unnecessary if-statements. But I'm not to strong on my decision here, I just need a way to associate the source code's type annotation.